File: frontend/src/lib/cdn.ts
import { env } from '$env/dynamic/public';

const BASE = env.PUBLIC_CDN_BASE_URL ?? '/cdn';

export type Transform = {
  w?: number;
  h?: number;
  fit?: 'cover' | 'contain';
  q?: number;
  fm?: 'auto' | 'jpg' | 'webp';
};

export function cdn(photoId: string, t: Transform = {}): string {
  const url = new URL(`${BASE}/img/${photoId}`, 'http://placeholder');
  if (t.w)   url.searchParams.set('w',  String(t.w));
  if (t.h)   url.searchParams.set('h',  String(t.h));
  if (t.fit) url.searchParams.set('fit', t.fit);
  if (t.q)   url.searchParams.set('q',  String(t.q));
  if (t.fm)  url.searchParams.set('fm', t.fm);
  // strip placeholder origin
  return url.pathname + url.search;
}

export function srcset(photoId: string, widths: number[], t: Omit<Transform, 'w'> = {}): string {
  return widths.map(w => `${cdn(photoId, { ...t, w })} ${w}w`).join(', ');
}

File: backend/src/photos/cdn.rs
//! CDN URL builder. Returns
//!   {base}/img/{photo_id}?w=&h=&fit=&q=&fm=
//! In prod, base is the CloudFront distribution. In dev, the backend's
//! own /cdn route serves the same URL shape locally.

use uuid::Uuid;

#[derive(Default)]
pub struct Transform {
    pub w: Option<u32>,
    pub h: Option<u32>,
    pub fit: Option<&'static str>, // "cover" | "contain"
    pub q: Option<u8>,
    pub fm: Option<&'static str>, // "auto" | "jpg" | "webp"
}

pub fn url(base: &str, photo_id: Uuid, t: &Transform) -> String {
    let mut s = format!("{base}/img/{photo_id}");
    let mut sep = '?';
    if let Some(w) = t.w {
        s.push(sep);
        s.push_str(&format!("w={w}"));
        sep = '&';
    }
    if let Some(h) = t.h {
        s.push(sep);
        s.push_str(&format!("h={h}"));
        sep = '&';
    }
    if let Some(f) = t.fit {
        s.push(sep);
        s.push_str(&format!("fit={f}"));
        sep = '&';
    }
    if let Some(q) = t.q {
        s.push(sep);
        s.push_str(&format!("q={q}"));
        sep = '&';
    }
    if let Some(fm) = t.fm {
        s.push(sep);
        s.push_str(&format!("fm={fm}"));
    }
    s
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn no_params() {
        let id = Uuid::nil();
        assert_eq!(
            url("https://cdn.x.app", id, &Transform::default()),
            format!("https://cdn.x.app/img/{id}")
        );
    }

    #[test]
    fn full_params() {
        let id = Uuid::nil();
        let t = Transform {
            w: Some(800),
            h: Some(600),
            fit: Some("cover"),
            q: Some(85),
            fm: Some("auto"),
        };
        let got = url("https://cdn.x.app", id, &t);
        assert!(got.contains("w=800"));
        assert!(got.contains("fm=auto"));
    }
}
